README.md
# @fanfic/proxy

This is a pretty basic HTTP proxy, of a sort. 

# FEATURES

* Rate limiting: System-wide and per domain
* Concurrency limiting: System-wide and per domain
* Per domain cookie injection
* Retries with back off on timeouts, errors.
* Compliant retries of 429s
* Detailed visibility into status via an HTTP scoreboard (HTML and JSON).
* Ability to force requests to come from a cache, for working with a snapshot.

# LIMITATIONS

* The cache does not follow HTTP cache semantics.  All responses are stored
  in the cache.  Requests get the fresh version unless they explicitly
  request the cached version.  Cache headers from the response are currently
  ignored.
* While it can be used as a standard proxy for HTTP requests, use with
  HTTPS either requires the requestor send `GET https://...` type queries
  (instead of `CONNECT`), or that it use a non-standard request format
  (`GET /https://...`).  

# USAGE

```console
$ npx @fanfic/proxy config.toml
```

Alternatively, install it and run it:

```console
$ npm i @fanfic/proxy
$ npx proxy config.toml
```

Or use it as a library:

```
const startProxy = require('@fanfic/proxy')
const config = { ... }
startProxy(config).then(() => {
  console.log('Proxy closed.')
}).catch(err => {
  console.error('Proxy errored:', err)
})
```

# WHY THE NAME?

Ok, so it's not particularly fanfiction specific, but if you're scraping
fanfic, needing to keep under site "you're spidering us too hard" thresholds
is super important, and this is a tool specifically built to do that.

Also, the ability to inject auth is super super useful--many sites put
content behind auth-walls and this makes spidering it much easier, not
requring that knowledge to live in your spider.

Could this be done with a real proper http proxy? Sure, but setting this up
is distinctly simpler.

Also the ability to force use of the cache is helpful in development
when you may need to rerequest the same page 50 times.

# USING THE PROXY

Using the proxy is as easy as sticking the URL you want on the end of the
URL for the proxy server.

```javascript
const fetch = require('node-fetch')
const proxyServer = 'http://localhost:10700'
const link = 'https://example.com'
fetch(`${proxyServer}/${link}`, opts)
```

## REQUESTING FROM CACHE

```javascript
// To fetch ONLY from the cache (and 504 if not in the cache):
fetch(`${proxyServer}/${link}`, {headers: {cache-control: 'only-if-cached'}})

// To fetch from the cache if available (but hit the network if not):
fetch(`${proxyServer}/${link}`, {headers: {cache-control: 'prefer-cached'}})
```

## ALL TOGETHER

So for example, to get a fetch that will send all requests to the proxy
server and prefer cached versions:

```javascript
const nodeFetch = require('node-fetch')
const proxyServer = 'http://localhost:10700'
const fetch = (link, opts = {}) =>
  nodeFetch(`${proxyServer}/${link}`, {'cache-control': 'prefer-cached', ...opts})
```

# VIEWING THE STATUS

If the proxy was running on port 10700 then visit `http://localhost:10700/`
and you'll get summary of the current status of the proxy.

# EXAMPLE CONFIG

```toml
[agent]
name = "ExampleProxy"
version = "1"
homepage = "https://example.com"

[proxy]
port = 10700
requestlog = true
cache = true
cachedir = 'cache/'

[global.limits]
per-second = 120
concurrent = 60

[sites.default]
limits = { per-second = 8, concurrent = 4, retries = 5 }

[sites."very-limited.example.com"]
limits = { per-second = 1, concurrent = 2 }

[sites."needs-authentication.example.com"]
cookies = [
  "auth_cookie=abc-def-ghi-jkl-ghi-123-456-789"
]

```

# CONFIG

## templatedir = <string>

Place to look for the scoreboard template file `proxy-scoreboard.html`. 
Defaults to the one bundled with the module.

## verify = <boolean>

Default: true. If true and caching is enabled, the cache will be verified and stale entries removed at startup time. With a large cache this may make a goodly number of seconds.

## agent = <object>

Configures how the proxy represents itself to the world.

### agent.name

Required. The name of your proxy, will show up in the user-agent.

### agent.version

Required. The version of your proxy, will show up in the user agent

### agent.homepage

Required. This should be a web page describing what you're doing and how to contact you if your spidering is causing problems for a site. This shows up in the user agent.

### agent.maxSockets

Default: 60.  This needs to be >= the global concurrent limit.  This is
per-protocol so it would allow 60 http and 60 https sockets.

### agent.outboundAddress

## proxy = <object>

General configuration about the service itself

### proxy.port = <number>

Default: 10700. The port for the proxy to listen on.

### proxy.requestlog = <boolean>

Default: true. Print a request log to STDOUT.

### proxy.cache = <boolean>

Default: true. Record requests to a cache.

### proxy.cachedir = <string>

Default: 'cache/'.  The directory to cache requests in.  Defaults to cache/
under the current location.

## global = <object>

Stores global request limits that are computed across ALL requests
regardless of domain

### global.limits = {per-second, concurrent, minimum-gap}

Limits across all sites, combined.  Used to protect you and your network
network, not the site being scraped.

### global.limits.per-second = <number>

Default: Infinity. Maximum number of requests per second to make.

### global.limits.concurrent = <number>

Default: Infinity. Maximum number of current requests allowed at a time.

### global.limits.minimum-gap = <number>

Default: 0. Minimum time that must elapse between a request completing and a new request being made.

## sites = <object>

This is a map of domain names to objects configuring requests to that
domain.

### DEFAULT SITE

There is a special site named `default` which is used to set default limits.

### SITE PROPERTIES

#### site.cookies = []

This is an array of cookies to be sent to a site, it should look like:

```
  "cookie_name=cookie-value"
```

You can restrict cookies to only part of the site by adding a `Path` component:

```
  "cookie_name=cookie-value; Path=/example/"
```

#### site.limits = {per-second, concurrent, minimum-gap, retries}

Site limits are like global limits

#### site.limits.per-second = <number>

Default: 8. Maximum number of requests per second to make.

#### site.limits.concurrent = <number>

Default: 4. Maximum number of current requests allowed at a time.

#### site.limits.minimum-gap = <number>

Default: 0.  Minimum time that must elapse between a request completing and
a new request being made.

#### site.limits.retries = <number>

Default: 5.  Maximum number of times to retry a failing request.  Requests
can fail due to timeouts, server errors, or explit requests to query more
slowly (eg, the server sending a 429).

#### site.limits.timeout = <number>

Default: 15. Seconds in which a request must complete. 

# RETRY DELAYS

When a retry is needed, the retry is delayed for an amount of time computed
based on this formula:

```
1.5 + (0.5 * (tries ** 2))
```

This exponential backoff means that if the server is overloaded we'll give
it more and more space to get things together before trying again.

The first retry will sleep 2 seconds before trying again and the 5th will
sleep 14 seconds.  This also means that if you were to bump the number of
retries out to 10, it would be sleeping 51.5 seconds before making that
final request.

# LIMITS

So we have three sorts of limits: concurrent, per-second and minimum-gap.

The reason for this is because of the different ways different servers
determine what constitutes too much load.  The idea is to tune these to
match with how the server you're talking to defines its limits.

## concurrent

Web browsers set this limit as between 4 and 8, so if you're in that range
most sites won't get mad at you.  However, it's worth noting that real world
use means that most users have a single main page request, and the rest are
static content--scripts and images, which are lower resource consumption for
most services.  Servers that do implement connection limits usually do not
do it in terms of concurrency.

## per-second

This is the maximum number of requests per second to make.  This is based on
when the request was started, so you may end up with more concurrent
requests than you allow per second, because some of them were started
earlier. This is the most common restriction on requests.

## minimum-gap

The minimum number of seconds between completed requests.  This is useful
when your requests take a long time to process--it ensures that the target
server gets some breathing room between your requests.

# TODO

Tests.  Some tests would be nice.  This is also the kind of project that's
the most painful to test.  Thus the lack of them.  I make practical use of
it quite a lot, however.

This pretty much does everything I need it to do, but I would welcome
patches that add the following:

* Proper HTTP cache support--using cache control headers from the servers
  and submitting requests with etags and if-modified-since.
* A little client library wrapper.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# 1.0.0 (2019-05-03)


fetch-back-off.js
'use strict'
const moment = require('moment')
const url = require('url')
const { URL } = require('url')
const { canonicalDomain, CookieJar } = require('tough-cookie')
const fun = require('funstream')
const rethrow = require('./rethrow.js')

function sleep (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}
async function timeout (ms) {
  const err = new Error('timeout')
  err.name = 'Timeout'
  err.code = 'ETIMEOUT'
  Error.captureStackTrace(err, timeout)
  await sleep(ms)
  throw err
}

let queueRunner
const state = {
  hosts: {},
  inflight: {},
  enqueued: 0,
  flying: 0,
  nextReq: null,
  lastReq: null,
  lastComplete: null,
  config: null
}

module.exports = (fetch, _config) => {
  const config = Object.assign({sites: {}}, _config || {})
  if (!config.global) config.global = {}
  if (!config.global.limits) config.global.limits = {}
  if (!config.sites.default) config.sites.default = {}
  state.config = config
  for (let name of Object.keys(config.sites)) {
    if (name === 'default') continue
    const site = config.sites[name]
    if (!site.cookies) continue
    site.cookieJar = new CookieJar()
    for (let cookie of site.cookies) {
      site.cookieJar.setCookieSync(`${cookie}; Domain=${name}`, `http://${name}`)
    }
    delete site.cookies
  }
  async function backoffFetch (uri, opts) {
    if (state.inflight[uri]) return state.inflight[uri]
    const site = siteConfig(uri, config)
    if (!site.cookieJar) site.cookieJar = new CookieJar()
    state.inflight[uri] = enqueue(fetch, uri, {cookieJar: site.cookieJar, ...opts})
    try {
      return await state.inflight[uri]
    } finally {
      delete state.inflight[uri]
    }
  }
  backoffFetch.defaults = fetch.defaults

  return backoffFetch
}
module.exports.state = state

function enqueue (fetch, uri, opts) {
  const info = hostState(uri)
  ++state.enqueued
  return new Promise(resolve => {
    info.queue.push({
      fetch,
      uri,
      opts,
      done: _ => {
        if (--state.enqueued === 0) stopRunner()
        return resolve(_)
      }
    })
    startRunner()
  })
}

function startRunner () {
  if (queueRunner) return
  queueRunner = setInterval(runQueue, 150)
  runQueue()
}

function stopRunner () {
  if (!queueRunner) return
  clearInterval(queueRunner)
  queueRunner = null
}

function hostname (uri) {
  return new URL(uri).host
}

function hostState (uri) {
  const host = hostname(uri)
  if (!state.hosts[host]) state.hosts[host] = {name: host, queue: [], flying: 0, nextReq: null, lastReq: null, lastComplete: null, lastErr: null}
  return state.hosts[host]
}

let last = Number(moment())
async function runQueue () {
  const now = Number(moment())

  if (state.nextReq && state.nextReq > now) return

  if (state.config.global.limits.concurrent && (state.flying >= state.config.global.limits.concurrent)) {
    process.emit('warn', 'Global: Defering due to global concurrency controls', state.flying)
    return
  }

  if (state.config.global.limits['per-second']) {
    const secondsPerRequest = 1 / state.config.global.limits['per-second']
    const sinceLast = state.lastReq ? (Number(moment()) - state.lastReq) / 1000 : Infinity
    const secondsBetweenRequests = state.config.global.limits['minimum-gap'] || 0
    const sinceLastComplete = state.lastComplete ? (Number(moment()) - state.lastComplete) / 1000 : Infinity
    if (sinceLast < secondsPerRequest) {
      process.emit('warn', `Global: Delaying for ${secondsPerRequest - sinceLast}s due to rate controls (${secondsPerRequest} versus ${sinceLast}`)
      const next = now + ((secondsPerRequest - sinceLast) * 1000)
      if (!state.nextReq || next > state.nextReq) state.nextReq = next
      return
    }
    if (sinceLastComplete < secondsBetweenRequests) {
      process.emit('warn', `Global: Delaying for ${secondsBetweenRequests - sinceLastComplete}s due to rate completion controls`)
      const next = now + ((secondsBetweenRequests - sinceLastComplete) * 1000)
      if (!state.nextReq || next > state.nextReq) state.nextReq = next
      return
    }
  }

  for (let name of Object.keys(state.hosts)) {
    const host = state.hosts[name]
    if (host.nextReq && host.nextReq > now) continue
    host.lastErr = null
    while (host.queue.length) {
      const info = host.queue[0]
      const perSiteLimits = {...defaultLimits(state.config), ...siteLimits(info.uri, state.config)}
      const maxRetries = perSiteLimits['retries'] || 5
      if (host.flying >= (perSiteLimits['concurrent'] || 4)) {
        process.emit('warn', 'Defering due to concurrency controls', host.name, host.flying)
        host.lastErr = 'Limit: Concurrency'
        break
      }
      const secondsPerRequest = 1 / (perSiteLimits['per-second'] || 8)
      const sinceLast = host.lastReq ? (Number(moment()) - host.lastReq) / 1000 : Infinity
      const secondsBetweenRequests = perSiteLimits['minimum-gap'] || 0
      const sinceLastComplete = host.lastComplete ? (Number(moment()) - host.lastComplete) / 1000 : Infinity
      if (sinceLast < secondsPerRequest) {
        process.emit('warn', `Delaying for ${secondsPerRequest - sinceLast}s due to rate controls (${secondsPerRequest} versus ${sinceLast}`)
        const next = now + ((secondsPerRequest - sinceLast) * 1000)
        host.nextReq = next
        host.lastErr = 'Limit: Request Rate'
        break
      }
      if (sinceLastComplete < secondsBetweenRequests) {
        process.emit('warn', `Delaying for ${secondsBetweenRequests - sinceLastComplete}s due to rate completion controls`)
        const next = now + ((secondsBetweenRequests - sinceLastComplete) * 1000)
        host.nextReq = next
        host.lastErr = 'Limit: Completion Rate'
        break
      }
      host.queue.shift()
      host.lastReq = now
      if (!state.lastReq || last > state.lastReq) state.lastReq = now
      ++host.flying
      ++state.flying
      const fetchOpts = {...info.opts}
      const cj = info.opts.cookieJar
      if (cj) {
        if (!fetchOpts.headers) fetchOpts.headers = {}
        fetchOpts.headers['Cookie'] = cj.getCookieStringSync(info.uri)
      }
      let res, err
      try {
        res = await Promise.race([timeout((perSiteLimits['timeout'] * 1000) || 15000), info.fetch(info.uri, fetchOpts)])
        if (res.headers && res.headers.has('set-cookie')) {
          for (let rawCookie of res.headers.raw()['set-cookie']) {
            try {
              await setCookieP(info.opts.cookieJar, rawCookie, res.url || info.uri)
            } catch (_) { /* ignore */ }
          }
        }
      } catch (_) {
        err = _
      }
      --host.flying
      --state.flying
      state.lastComplete = host.lastComplete = Number(moment())
      if (!err && res.status < 400) {
        info.done(res)
        return
      }
      if (!info.tries) info.tries = 0
      ++info.tries
      if (info.tries > maxRetries) return info.done(Promise.reject(rethrow(err) || resToError(res, info.uri) || new Error(`Ran out of retries ${info.tries} > ${maxRetries}`)))
      const retryDelay = 1500 + (500 * (info.tries ** 2))
      if (res && res.status >= 500) {
        process.emit('warn', `Server error on ${info.uri} sleeping`, retryDelay / 1000, 'seconds')
        host.queue.unshift(info)
        host.nextReq = Number(moment()) + retryDelay
        host.lastErr = 'Error: Server Error'
        if (res && res.body) await fun(res.body).forEach(_ => {}).catch(_ => {})
      } else if ((res && res.status === 408) || (err && (err.code === 'ETIMEOUT' || err.type === 'body-timeout' || /timeout/i.test(err.message)))) {
        process.emit('warn', `Timeout on ${info.uri} sleeping`, retryDelay / 1000, 'seconds')
        host.queue.unshift(info)
        host.nextReq = Number(moment()) + retryDelay
        host.lastErr = 'Error: Timeout'
        if (res && res.body) await fun(res.body).forEach(_ => {}).catch(_ => {})
      } else if (res && res.status === 429) {
        const retryAfter = res.headers['retry-after']
        let retryTime = 3000 + (500 * (info.tries ** 2))
        if (retryAfter) {
          if (/^\d+$/.test(retryAfter)) {
            retryTime = Number(retryAfter) * 1000
          } else {
            retryTime = (moment().unix() - moment.utc(retryAfter, 'ddd, DD MMM YYYY HH:mm:ss ZZ').unix()) * 1000
          }
        }
        process.emit('warn', 'Request backoff requested, sleeping', retryTime / 1000, 'seconds', `(${retryAfter ? 'at: ' + retryAfter + ', ' : ''}now: ${moment.utc()})`)
        host.queue.unshift(info)
        host.nextReq = Number(moment()) + retryTime
        host.lastErr = 'Error: 429 Too Many Requests'
        if (res && res.body) await fun(res.body).forEach(_ => {}).catch(_ => {})
      } else if (err) {
        if (res && res.body) await fun(res.body).forEach(_ => {}).catch(_ => {})
        if (err.code === 'HPE_HEADER_OVERFLOW') {
          info.done({
            status: 502,
            statusText: 'Reponse headers too large',
            body: fun('<h1>Response headers were &gt; 8kb, which is not supported by Node.js</h1>'),
            headers: new Map([
              ['Content-Type', 'text/html']
            ])
          })
        } else {
          info.done(Promise.reject(rethrow(err)))
        }
      } else {
        info.done(res)
      }
    }
  }
}

function siteConfig (uri, conf) {
  if (!conf) conf = {}
  if (!conf.sites) conf.sites = {}
  const hostname = canonicalDomain(new URL(uri).hostname)
  const domain = hostname.split(/[.]/)
  while (domain.length > 0) {
    const host = domain.join('.')
    if (host in conf.sites) return conf.sites[host]
    domain.shift()
  }
  return conf.sites[hostname] = {}
}

function defaultConfig (conf) {
  return (conf && conf.sites && conf.sites.default) || {}
}

function siteLimits (uri, conf) {
  return siteConfig(uri, conf).limits || {}
}
function defaultLimits (uri, conf) {
  return defaultConfig(uri, conf).limits || {}
}

function setCookieP (jar, cookie, link) {
  const linkP = new URL(link)
  linkP.pathname = ''
  return new Promise((resolve, reject) => {
    jar.setCookie(cookie, url.format(linkP), (err, cookie) => {
      return err ? reject(rethrow(err)) : resolve(cookie)
    })
  })
}

function resToError (meta, toFetch) {
  if (!meta) return
  if (meta.status === 403) {
    const err = new Error('Got status: ' + meta.status + ' ' + meta.statusText + ' for ' + toFetch)
    Error.captureStackTrace(err, resToError)
    err.code = meta.status
    err.url = toFetch
    err.meta = meta
    return err
  } else if (meta.status === 429) {
    const err = new Error('Got status: ' + meta.status + ' ' + meta.statusText + ' for ' + toFetch)
    Error.captureStackTrace(err, resToError)
    err.code = meta.status
    err.url = toFetch
    err.meta = meta
    err.retryAfter = meta.headers['retry-after']
    return err
  } else if (meta.status === 404) {
    const err = new Error('Got status: ' + meta.status + ' ' + meta.statusText + ' for ' + toFetch)
    Error.captureStackTrace(err, resToError)
    err.code = meta.status
    err.url = toFetch
    err.meta = meta
    return err
  } else if (meta.status && (meta.status < 200 || meta.status >= 400)) {
    const err = new Error('Got status: ' + meta.status + ' ' + meta.statusText + ' for ' + toFetch)
    Error.captureStackTrace(err, resToError)
    err.code = meta.status
    err.url = toFetch
    err.meta = meta
    return err
  }
}


proxy.js
'use strict'
const fs = require('fs')
const url = require('url')
const qr = require('@perl/qr')
const Koa = require('koa')
const cacache = require('cacache/en')
const plainFetch = require('node-fetch')
const fetchBackOff = require('./fetch-back-off.js')
const logger = require('koa-logger')
const http = require('http')
const HttpAgent = require('agentkeepalive')
const HttpsAgent = require('agentkeepalive').HttpsAgent
const fun = require('funstream')
const Gzip = require('minizlib').Gzip
const Gunzip = require('minizlib').Gunzip
const Handlebars = require('handlebars')
const moment = require('moment')

function makeCacache (cachedir) {
  return {
    verify: () => cacache.verify(cachedir),
    getInfo: key => cacache.get.info(cachedir, key),
    getStream: key => cacache.get.stream(cachedir, key),
    rmEntry: key => cacache.rm.entry(cachedir, key),
    putStream: (key, options) => cacache.put.stream(cachedir, key, options)
  }
}

module.exports = async userConf => {
  if (!userConf) userConf = {}
  if (!userConf.proxy) userConf.proxy = {}
  if (!userConf.agent) userConf.agent = {}
  const conf = {
    ...userConf,
    proxy: {
      port: 10700,
      cache: true,
      cachedir: `cache/`,
      requestlog: true,
      ...userConf.proxy
    },
    agent: {
      maxSockets: 60,
      strictSSL: true,
      ...userConf.agent
    }
  }

  if (!conf.agent.name) throw new Error('Must provide agent.name config value, how to identify your app to sites you scrape from')
  if (!conf.agent.version) throw new Error('Must provide agent.version config value, the version of your app')
  if (!conf.agent.homepage) throw new Error('Must provide agent.homepage config value, a page describing what your app does and how to contact you')

  conf.fetch = fetchBackOff(plainFetch, conf)
  conf.cache = conf.proxy.cache && makeCacache(conf.proxy.cachedir)

  const templateDir = conf.templatedir || __dirname
  /* eslint-disable security/detect-non-literal-fs-filename */
  const sbTemplate = Handlebars.compile(fs.readFileSync(`${templateDir}/proxy-scoreboard.html`, 'utf8'))

  conf.agent['http:'] = new HttpAgent({
    maxSockets: conf.agent.maxSockets,
    localAddress: conf.agent.outboundAddress
  })
  conf.agent['https:'] = new HttpsAgent({
    maxSockets: conf.agent.maxSockets,
    ca: conf.agent.ca,
    cert: conf.agent.cert,
    key: conf.agent.key,
    localAddress: conf.agent.outboundAddress,
    rejectUnauthorized: conf.agent.strictSSL
  })

  if (conf.proxy.cache && conf.verify) {
    console.log('verifying cache...')
    await conf.cache.verify()
  }

  console.log('firing up service')
  const proxy = new Koa()
  proxy.use(fixupHttps)
  if (conf.proxy.requestlog) proxy.use(logger())
  proxy.use(routeRequest)
  const proxySrv = http.createServer(proxy.callback()).listen(conf.proxy.port, () => console.error('listening'))

  await new Promise((resolve, reject) => {
    proxy.on('error', reject)
    proxySrv.on('error', reject)
    process.on('SIGINT', resolve)
  })
  proxySrv.close()

  async function fixupHttps (ctx, next) {
    const reqUrl = url.parse(ctx.url)
    if (reqUrl.protocol === 'https:' || reqUrl.port === 443) {
      reqUrl.protocol = 'https:'
      delete reqUrl.port
      reqUrl.host = reqUrl.hostname
      ctx.url = url.format(reqUrl)
    }
    if (/^[/]https?:/.test(reqUrl.path)) {
      ctx.url = reqUrl.path.slice(1)
    }
    return next()
  }

  async function routeRequest (ctx, next) {
    try {
      if (ctx.url === '/') {
        const accepts = ctx.accepts('html', 'json')
        if (accepts === 'json') {
          ctx.status = 200
          ctx.body = scoreBoard()
        } else {
          ctx.status = 200
          ctx.body = sbTemplate(scoreBoard())
        }
        return next()
      } else if (qr`^https?://`.test(ctx.url)) {
        ctx.body = fun()
          .on('error', err => {
            ctx.throw(500, err.message)
          })
        const hasCache = conf.proxy.cache
        const onlyIfCached = ctx.header['cache-control'].toLowerCase() === 'only-if-cached'
        const preferCached = ctx.header['cache-control'].toLowerCase() === 'prefer-cached'
        if (hasCache && (onlyIfCached || preferCached)) {
          try {
            return await fromCache(ctx, conf, next)
          } catch (_) {
            // ignored because we'll just try the network instead
          }
        }
        // TODO: Allow more typical cache behavior, eg check cache freshness, look up last modified times for
        // if-modified-since checking, etc. Low priority, however, given the core usecase of this module.
        if (onlyIfCached) {
          ctx.throw(504, 'Not in cache')
          return
        }
        return await fromNetwork(ctx, conf, next)
      }
    } catch (err) {
      console.error('fromRequest error', err)
      ctx.throw(500, 'errored')
    }
  }
}

async function fromCache (ctx, conf, next) {
  const {metadata} = await conf.cache.getInfo(ctx.url)
  let errored = false
  const cacheResult = fun(conf.cache.getStream(ctx.url))
    .pipe(new Gunzip())
    .once('error', async err => {
      if (errored) {
        if (errored !== err.message) console.error(err.message)
        return
      }
      errored = err.message
      await conf.cache.rmEntry(ctx.url).catch(_ => {})
      fromNetwork(ctx, conf, next)
    })
    .on('data', async function onData (data) {
      cacheResult.removeListener('data', onData)
      const status = metadata.status
      if (status === 429) {
        await conf.cache.rmEntry(ctx.url).catch(_ => {})
        return fromNetwork(ctx, conf, next)
      }
      if (status) ctx.status = status
      const headers = metadata.headers ? metadata.headers : metadata
      for (let header of Object.entries(headers)) {
        const [key, value] = header
        ctx.set(key, value)
      }
      ctx.body.write(data)
      cacheResult.pipe(ctx.body)
      next()
    })
}

async function fromNetwork (ctx, conf, next) {
  const userAgent = `Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:59.0; ${conf.agent.name}/${conf.agent.version}; +${conf.agent.homepage}) Gecko/20100101 Firefox/59.0`
  const reqUrl = url.parse(ctx.url)
  const agent = conf.agent[reqUrl.protocol]
  const reqHeaders = {}
  reqHeaders['User-Agent'] = userAgent
  if (ctx.header.accept) reqHeaders['Accept'] = ctx.header.accept
  if (ctx.header['accept-encoding']) reqHeaders['Accept-Encoding'] = ctx.header['accept-encoding']
  if (ctx.header['content-type']) reqHeaders['Content-Type'] = ctx.header['content-type']
  if (ctx.header['content-length']) reqHeaders['Content-Length'] = ctx.header['content-length']
  const opts = {
    headers: reqHeaders,
    method: ctx.method,
    agent,
    body: (ctx.method === 'POST' || ctx.method === 'PUT') ? ctx.req : undefined,
    redirect: 'manual'
  }
  if (opts.body) opts.body.on('error', err => ctx.throw(500, err.message))
  try {
    const response = await conf.fetch(ctx.url, opts)
    ctx.status = response.status
    const body = response.body
    const headers = {}
    for (let header of response.headers.entries()) {
      const [key, value] = header
      if (key === 'transfer-encoding' || key === 'content-encoding' || key === 'content-length' || key === 'connection') continue
      headers[key] = value
      ctx.set(key, value)
    }
    ctx.set('FF-Final-URL', response.finalUrl || ctx.url)
    body.pipe(ctx.body)
      .on('error', err => console.error('body->response error', err))

    if (conf.proxy.cache) {
      // errors from cache storage are ignored because things work anyway
      body.pipe(new Gzip())
        .on('error', err => console.error('body->gzip error', err))
        .pipe(conf.cache.putStream(ctx.url, {metadata: {status: response.status, headers}}))
        .on('error', err => console.error('gzip->cache error', err))
    }
    next()
  } catch (err) {
    ctx.status = 500
    console.error('fromNetwork error', err)
    return next()
  }
}

function formatDate (date) {
  if (!date) return '-'
  return moment(date).toString()
}

function scoreBoard () {
  const state = fetchBackOff.state
  const sb = {
    inflightCount: Object.keys(state.inflight).length,
    queued: state.enqueued,
    flying: state.flying,
    nextReq: formatDate(state.nextReq),
    lastReq: formatDate(state.lastReq),
    lastComplete: formatDate(state.lastComplete),
    hosts: []
  }
  for (let host of Object.values(state.hosts)) {
    sb.hosts.push({
      name: host.name,
      queued: host.queue.length,
      flying: host.flying,
      nextReq: formatDate(host.nextReq),
      lastReq: formatDate(host.lastReq),
      lastComplete: formatDate(host.lastComplete),
      lastErr: host.lastErr || ''
    })
  }
  return sb
}
